File: sdk/python/quine_rs/websocket.py
"""
WebSocket subscription client for Nexora-RS Standing Query results.

Connects to the ``/api/v2/ws/sq/{id}`` WebSocket endpoint to receive
real-time notifications when a Standing Query matches.

Example::

    import asyncio
    from nexora_rs import AsyncNexoraClient, StandingQuerySubscription

    async def main():
        async with AsyncNexoraClient("http://localhost:8080") as client:
            sq_id = await client.create_standing_query(
                {"type": "PropertyFilter", "key": "speed",
                 "condition": {"type": "GreaterThan", "value": 100}},
                name="fast_cars",
            )

            ws_url = "ws://localhost:8080"
            async with StandingQuerySubscription(ws_url, sq_id) as sub:
                async def on_result(data):
                    print(f"Match! count={data['match_count']}")

                await sub.subscribe(on_result)

    asyncio.run(main())
"""

from __future__ import annotations

import asyncio
import json
import logging
from typing import Any, Awaitable, Callable

import aiohttp

from .exceptions import (
    AuthenticationError,
    ConnectionError as NexoraConnectionError,
    NexoraError,
    TimeoutError as NexoraTimeoutError,
)

logger = logging.getLogger(__name__)

# Type alias for the async callback
ResultCallback = Callable[[dict[str, Any]], Awaitable[None]]


class StandingQuerySubscription:
    """Subscribe to Standing Query results via WebSocket.

    Args:
        ws_url: WebSocket base URL (e.g. ``ws://localhost:8080``).
        query_id: The Standing Query UUID to subscribe to.
        api_key: Optional Bearer token for authentication.
    """

    def __init__(
        self,
        ws_url: str,
        query_id: str,
        api_key: str | None = None,
    ) -> None:
        self.ws_url = ws_url.rstrip("/")
        self.query_id = query_id
        self.api_key = api_key
        self._session: aiohttp.ClientSession | None = None
        self._ws: aiohttp.ClientWebSocketResponse | None = None
        self._running = False

    # ------------------------------------------------------------------
    # Context manager
    # ------------------------------------------------------------------

    async def __aenter__(self) -> StandingQuerySubscription:
        await self._connect()
        return self

    async def __aexit__(self, *args: Any) -> None:
        await self.unsubscribe()

    # ------------------------------------------------------------------
    # Connection
    # ------------------------------------------------------------------

    async def _connect(self) -> None:
        """Establish the WebSocket connection."""
        if self._session is None or self._session.closed:
            headers: dict[str, str] = {}
            if self.api_key:
                headers["Authorization"] = f"Bearer {self.api_key}"
            self._session = aiohttp.ClientSession()
        ws_url = f"{self.ws_url}/api/v2/ws/sq/{self.query_id}"
        try:
            self._ws = await self._session.ws_connect(ws_url, headers=headers)
        except aiohttp.ClientConnectorError as exc:
            raise NexoraConnectionError(
                f"Cannot connect to WebSocket {ws_url}: {exc}"
            ) from exc
        except aiohttp.WSServerHandshakeError as exc:
            if exc.status in (401, 403):
                raise AuthenticationError(
                    f"WebSocket auth failed: {exc}", status_code=exc.status
                ) from exc
            raise NexoraConnectionError(
                f"WebSocket handshake failed: {exc}"
            ) from exc

    # ------------------------------------------------------------------
    # Subscribe / Unsubscribe
    # ------------------------------------------------------------------

    async def subscribe(self, on_result: ResultCallback) -> None:
        """Start receiving Standing Query results.

        Calls ``on_result`` for each message received from the server.
        This method blocks until :meth:`unsubscribe` is called, the
        connection closes, or an error occurs.

        Args:
            on_result: Async callback invoked with each result dict.
        """
        if self._ws is None:
            await self._connect()

        self._running = True
        assert self._ws is not None

        try:
            async for msg in self._ws:
                if not self._running:
                    break

                if msg.type == aiohttp.WSMsgType.TEXT:
                    try:
                        data = json.loads(msg.data)
                    except json.JSONDecodeError:
                        logger.warning("Received non-JSON WebSocket message: %s", msg.data)
                        continue

                    try:
                        await on_result(data)
                    except Exception:
                        logger.exception("Error in on_result callback")

                elif msg.type == aiohttp.WSMsgType.ERROR:
                    raise NexoraConnectionError(
                        f"WebSocket error: {self._ws.exception()}"
                    )
                elif msg.type in (aiohttp.WSMsgType.CLOSE, aiohttp.WSMsgType.CLOSING, aiohttp.WSMsgType.CLOSED):
                    break
        finally:
            self._running = False

    async def unsubscribe(self) -> None:
        """Stop receiving results and close the WebSocket connection."""
        self._running = False
        if self._ws and not self._ws.closed:
            await self._ws.close()
        if self._session and not self._session.closed:
            await self._session.close()
        self._ws = None
        self._session = None

    # ------------------------------------------------------------------
    # Utility
    # ------------------------------------------------------------------

    @property
    def is_running(self) -> bool:
        """Whether the subscription is currently active."""
        return self._running

File: sdk/typescript/src/websocket.ts
/**
 * WebSocket streaming clients for Nexora.
 *
 * Provides two subscription classes:
 *
 * - {@link StandingQuerySubscription} — receive real-time notifications
 *   when a Standing Query matches.
 * - {@link CypherStreamSubscription} — stream Cypher query results over
 *   a WebSocket connection.
 *
 * Both classes use the native `WebSocket` API (available in Node 18+ and
 * all modern browsers).
 *
 * @example Standing Query
 * ```ts
 * import { NexoraClient, StandingQuerySubscription } from 'nexora';
 *
 * const client = new NexoraClient('http://localhost:8080');
 * const sqId = await client.createStandingQuery(
 *   { type: 'PropertyFilter', key: 'speed',
 *     condition: { type: 'GreaterThan', value: 100 } },
 *   'fast_alert',
 * );
 *
 * const sub = new StandingQuerySubscription('ws://localhost:8080', sqId);
 * await sub.subscribe((data) => console.log('Match!', data));
 * // ... later
 * await sub.unsubscribe();
 * ```
 *
 * @example Cypher Stream
 * ```ts
 * const stream = new CypherStreamSubscription('ws://localhost:8080');
 * for await (const row of stream.stream('MATCH (n) RETURN n LIMIT 10')) {
 *   console.log(row);
 * }
 * ```
 */

import { ConnectionError } from './errors';

/** Callback type for receiving WebSocket messages. */
export type ResultCallback = (data: Record<string, unknown>) => void;

/**
 * Subscribe to Standing Query results via WebSocket.
 *
 * Connects to `/api/v2/ws/sq/{queryId}` and receives real-time
 * notifications whenever the Standing Query matches new data.
 */
export class StandingQuerySubscription {
  private readonly wsUrl: string;
  private readonly queryId: string;
  private readonly apiKey?: string;
  private ws: WebSocket | null = null;
  private running = false;

  constructor(wsUrl: string, queryId: string, apiKey?: string) {
    this.wsUrl = wsUrl.replace(/\/$/, '');
    this.queryId = queryId;
    this.apiKey = apiKey;
  }

  private buildUrl(): string {
    const base = `${this.wsUrl}/api/v2/ws/sq/${this.queryId}`;
    if (this.apiKey) {
      return `${base}?token=${encodeURIComponent(this.apiKey)}`;
    }
    return base;
  }

  /** Whether the subscription is currently active. */
  get isRunning(): boolean {
    return this.running;
  }

  /**
   * Start receiving Standing Query results.
   *
   * Calls `callback` for each message received from the server.
   * Resolves when the connection closes or an error occurs.
   */
  subscribe(callback: ResultCallback): Promise<void> {
    return new Promise((resolve, reject) => {
      this.running = true;
      const url = this.buildUrl();

      try {
        this.ws = new WebSocket(url);
      } catch (err) {
        this.running = false;
        reject(
          new ConnectionError(
            `Cannot create WebSocket: ${(err as Error).message}`,
          ),
        );
        return;
      }

      this.ws.onopen = () => {
        // Connection established
      };

      this.ws.onmessage = (event: MessageEvent) => {
        if (!this.running) return;
        const data = this.parseMessage(event.data);
        if (data) {
          try {
            callback(data);
          } catch (err) {
            console.error('Error in subscription callback:', err);
          }
        }
      };

      this.ws.onerror = () => {
        if (this.running) {
          this.running = false;
          reject(
            new ConnectionError(
              `WebSocket error for standing query ${this.queryId}`,
            ),
          );
        }
      };

      this.ws.onclose = () => {
        if (this.running) {
          this.running = false;
        }
        resolve();
      };
    });
  }

  /** Stop receiving results and close the WebSocket connection. */
  async unsubscribe(): Promise<void> {
    this.running = false;
    if (this.ws) {
      if (
        this.ws.readyState === WebSocket.OPEN ||
        this.ws.readyState === WebSocket.CONNECTING
      ) {
        this.ws.close();
      }
      this.ws = null;
    }
  }

  /**
   * Async iterator that yields each message from the server.
   *
   * @example
   * ```ts
   * const sub = new StandingQuerySubscription(url, id);
   * for await (const data of sub) {
   *   console.log(data);
   * }
   * ```
   */
  async *[Symbol.asyncIterator](): AsyncGenerator<Record<string, unknown>> {
    this.running = true;
    const url = this.buildUrl();

    const queue: Record<string, unknown>[] = [];
    let resolveWait: (() => void) | null = null;
    let rejectWait: ((err: Error) => void) | null = null;
    let done = false;

    try {
      this.ws = new WebSocket(url);
    } catch (err) {
      this.running = false;
      throw new ConnectionError(
        `Cannot create WebSocket: ${(err as Error).message}`,
      );
    }

    this.ws.onmessage = (event: MessageEvent) => {
      if (!this.running) return;
      const data = this.parseMessage(event.data);
      if (data) {
        queue.push(data);
        resolveWait?.();
      }
    };

    this.ws.onerror = () => {
      if (this.running) {
        done = true;
        this.running = false;
        rejectWait?.(
          new ConnectionError(
            `WebSocket error for standing query ${this.queryId}`,
          ),
        );
      }
    };

    this.ws.onclose = () => {
      done = true;
      this.running = false;
      resolveWait?.();
    };

    while (!done) {
      if (queue.length > 0) {
        yield queue.shift()!;
        continue;
      }

      await new Promise<void>((resolve, reject) => {
        resolveWait = resolve;
        rejectWait = reject;
      });
      resolveWait = null;
      rejectWait = null;
    }

    // Drain remaining
    while (queue.length > 0) {
      yield queue.shift()!;
    }
  }

  private parseMessage(data: unknown): Record<string, unknown> | null {
    if (typeof data !== 'string') return null;
    try {
      const parsed = JSON.parse(data);
      if (parsed !== null && typeof parsed === 'object' && !Array.isArray(parsed)) {
        return parsed as Record<string, unknown>;
      }
      return null;
    } catch {
      return null;
    }
  }
}

/**
 * Stream Cypher query results via WebSocket.
 *
 * Connects to `/api/v2/ws/query` and sends Cypher queries, receiving
 * results as streaming messages.
 */
export class CypherStreamSubscription {
  private readonly wsUrl: string;
  private readonly apiKey?: string;
  private ws: WebSocket | null = null;
  private running = false;

  constructor(wsUrl: string, apiKey?: string) {
    this.wsUrl = wsUrl.replace(/\/$/, '');
    this.apiKey = apiKey;
  }

  private buildUrl(): string {
    const base = `${this.wsUrl}/api/v2/ws/query`;
    if (this.apiKey) {
      return `${base}?token=${encodeURIComponent(this.apiKey)}`;
    }
    return base;
  }

  /** Whether the stream is currently active. */
  get isRunning(): boolean {
    return this.running;
  }

  /**
   * Start streaming Cypher query results.
   *
   * @param cypher   Cypher query string to execute.
   * @param callback Function invoked with each result message.
   * @param queryId  Optional client-side query identifier.
   */
  subscribe(
    cypher: string,
    callback: ResultCallback,
    queryId = 'q',
  ): Promise<void> {
    return new Promise((resolve, reject) => {
      this.running = true;
      const url = this.buildUrl();

      try {
        this.ws = new WebSocket(url);
      } catch (err) {
        this.running = false;
        reject(
          new ConnectionError(
            `Cannot create WebSocket: ${(err as Error).message}`,
          ),
        );
        return;
      }

      this.ws.onopen = () => {
        const request = JSON.stringify({ query: cypher, queryId });
        this.ws?.send(request);
      };

      this.ws.onmessage = (event: MessageEvent) => {
        if (!this.running) return;
        const data = this.parseMessage(event.data);
        if (!data) return;

        try {
          callback(data);
        } catch (err) {
          console.error('Error in stream callback:', err);
        }

        // Stop on QueryFinished
        if (data.type === 'QueryFinished') {
          this.running = false;
        }
      };

      this.ws.onerror = () => {
        if (this.running) {
          this.running = false;
          reject(new ConnectionError('WebSocket error for Cypher stream'));
        }
      };

      this.ws.onclose = () => {
        if (this.running) {
          this.running = false;
        }
        resolve();
      };
    });
  }

  /** Stop the stream and close the WebSocket connection. */
  async stop(): Promise<void> {
    this.running = false;
    if (this.ws) {
      if (
        this.ws.readyState === WebSocket.OPEN ||
        this.ws.readyState === WebSocket.CONNECTING
      ) {
        this.ws.close();
      }
      this.ws = null;
    }
  }

  /**
   * Stream Cypher results as an async iterator.
   *
   * @param cypher  Cypher query string to execute.
   * @param queryId Optional client-side query identifier.
   * @yields Each result message from the server.
   */
  async *stream(
    cypher: string,
    queryId = 'q',
  ): AsyncGenerator<Record<string, unknown>> {
    this.running = true;
    const url = this.buildUrl();

    const queue: Record<string, unknown>[] = [];
    let resolveWait: (() => void) | null = null;
    let rejectWait: ((err: Error) => void) | null = null;
    let done = false;

    try {
      this.ws = new WebSocket(url);
    } catch (err) {
      this.running = false;
      throw new ConnectionError(
        `Cannot create WebSocket: ${(err as Error).message}`,
      );
    }

    this.ws.onopen = () => {
      const request = JSON.stringify({ query: cypher, queryId });
      this.ws?.send(request);
    };

    this.ws.onmessage = (event: MessageEvent) => {
      if (!this.running) return;
      const data = this.parseMessage(event.data);
      if (data) {
        queue.push(data);
        resolveWait?.();

        if (data.type === 'QueryFinished') {
          this.running = false;
        }
      }
    };

    this.ws.onerror = () => {
      if (this.running) {
        done = true;
        this.running = false;
        rejectWait?.(new ConnectionError('WebSocket error for Cypher stream'));
      }
    };

    this.ws.onclose = () => {
      done = true;
      this.running = false;
      resolveWait?.();
    };

    while (!done) {
      if (queue.length > 0) {
        const msg = queue.shift()!;
        yield msg;
        if (msg.type === 'QueryFinished') break;
        continue;
      }

      await new Promise<void>((resolve, reject) => {
        resolveWait = resolve;
        rejectWait = reject;
      });
      resolveWait = null;
      rejectWait = null;
    }

    // Drain remaining
    while (queue.length > 0) {
      yield queue.shift()!;
    }
  }

  private parseMessage(data: unknown): Record<string, unknown> | null {
    if (typeof data !== 'string') return null;
    try {
      const parsed = JSON.parse(data);
      if (parsed !== null && typeof parsed === 'object' && !Array.isArray(parsed)) {
        return parsed as Record<string, unknown>;
      }
      return null;
    } catch {
      return null;
    }
  }
}
